File: TrabalhoParteFinal/Excecao.java
import java.io.File;
import java.io.FileWriter;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.InputMismatchException;
import java.util.Map;
import java.util.Scanner;


public class Excecao extends Exception {
    public Excecao(String msg) {

        super(msg);
    }

 

    public static class FinalizaProgramaException extends Excecao {
        public FinalizaProgramaException() {
            super("");
        }
    }

    public class ErroFormatacaoException extends Exception {
        public ErroFormatacaoException() {
            super("Erro de formatação.");
        }
    }

    public static class CodigosIguaisException extends Excecao {
        int codigo;

        public CodigosIguaisException(int codigo) {
            super("Código repetido para curso: " + codigo + ".");
            this.codigo = codigo;

        }

    }

    public static class MatriculasIguaisException extends Excecao {
        int matricula;

        public MatriculasIguaisException(int matricula) {
            super("Matrícula repetida para aluno: " + matricula + ".");
            this.matricula = matricula;

        }

    }

    public static class CodDisciplinaIndefinidoAlunoExcpetion extends Excecao {
        int matricula; // avaliacao ou aluno
        String codigo;

        public CodDisciplinaIndefinidoAlunoExcpetion(int matricula, String codigo) {
            super("código de didciplina não definido usado no aluno " + matricula + ": " + codigo);
            this.matricula = matricula;
            this.codigo = codigo;
        }
    }

    public static class CodDisciplinaIndefinidoAvalExcpetion extends Excecao {
        String codigoA; // avaliacao ou aluno
        String codigoD;

        public CodDisciplinaIndefinidoAvalExcpetion(String codigoA, String codigoD) {
            super("código de didciplina não definido usado na avaliação " + codigoA + ": " + codigoD);
            this.codigoA = codigoA;
            this.codigoD = codigoD;
        }
    }

    public static class PesoZeroNegativo extends Excecao {
        String codigo;
        double peso;

        public PesoZeroNegativo(String codigo, double peso) {
            super("Peso de avaliação inválido para " + codigo + ": " + peso);
            this.codigo = codigo;
            this.peso = peso;
        }
    }

    public static class NemPNemTException extends Excecao {
        String codigo;
        String Digitado;

        public NemPNemTException(String codigo, String digitado) {
            super("Tipo de avaliação desconhecido para " + codigo + ": " + digitado + ".");
            this.codigo = codigo;
            this.Digitado = digitado;
        }
    }

    public static class TamGrupoNaProvaException extends Excecao {
        String codigo;
        int tam;
        public TamGrupoNaProvaException(String codigo, int tam) {
            super("Tamanho máximo de grupo foi especificado a prova " + codigo +": " + tam);
        }
    }

    public static class TamMaxZeroNegativo extends Excecao {
        String codigo;
        double peso;

        public TamMaxZeroNegativo(String codigo, double peso) {
            super("Tamanho máximo de grupo invá1lido para trabalho " + codigo + ": " + peso);
            this.codigo = codigo;
            this.peso = peso;
        }
    }

    public static class NemGNemPException extends Excecao {
        int matricula;
        String Digitado;

        public NemGNemPException(int matricula, String digitado) {
            super("Tipo de aluno desconhecido para " + matricula + ": " + digitado + ".");
            this.matricula = matricula;
            this.Digitado = digitado;
        }
    }

    public static class CodCursoIndefinidoException extends Excecao{
        int matricula;
        int codigo;
        public CodCursoIndefinidoException(int matricula, int codigo) {
            super("Código de curso não definido usado no aluno " + matricula + ": " + codigo+".");
            this.matricula = matricula;
            this.codigo = codigo;
        }
    }

    public static class NemMNemDException extends Excecao {
        int matricula;
        String Digitado;

        public NemMNemDException(int  matricula, String digitado) {
            super("Tipo de aluno de pós-graduação desconhecido para " + matricula + ": " + digitado + ".");
            this.matricula = matricula;
            this.Digitado = digitado;
        }
    }

    public static class CodAvaliacaoIndefinidoException extends Excecao{
        ArrayList<Integer> M;
        String codigo;

        public CodAvaliacaoIndefinidoException(ArrayList<Integer> matriculas, String codigo){
            super("Códgo de avaliação não definido usado na planilha de notas, associado ao(s) aluno(s) ");
            for(int i=0;i<matriculas.size();i++){
                System.out.println(matriculas.get(i));
                if(i==matriculas.size()-1){
                    System.out.println(" ");
                }
            }
            System.out.println(".");
            this.M=matriculas;
            this.codigo=codigo;
        }
    }

    public static class MatriculaIndefinidaException extends Excecao{
        String codigo;
        int matricula;

        public MatriculaIndefinidaException(String codigo, int matricula){
            super("Matricula de aluno não definida usada na planilha de notas, asssociada à avaliação "+codigo+":"+matricula+".");
        }
    }

    public static class NotaInvalidaAvaliacaoException extends Excecao{
        ArrayList<Integer> M;
        String codigo;
        int nota;
        public NotaInvalidaAvaliacaoException(ArrayList<Integer> matriculas, String codigo, int nota){
            super("Nota inválida para avaliação "+ codigo + "do(s) aluno(s)");
            for(int i=0;i<matriculas.size();i++){
                System.out.println(matriculas.get(i));
                if(i==matriculas.size()-1){
                    System.out.println(" ");
                }
            }
            System.out.println(": "+nota+".");
        }
    }

    public static class DisciplinaSemAvaliacaoException extends Excecao{
        String codigo;
        public DisciplinaSemAvaliacaoException(String codigo){
            super("A disciplina "+ codigo+ "não possui nenhuma avaliação cadastrada.");
        }
    }

    public static class AlunoNaoMatriculadoException extends Excecao{
        int matricula;
        String codigoA;
        String codigoD;

        public AlunoNaoMatriculadoException(int matricula, String codigoA, String codigoD){
            super("O aluno "+ matricula+ "não possui nota na avalialiação "+codigoA+ "da disciplina "+ codigoD+"porém não encontra-se matriculado nesta disciplina.");

    }
}

    public static class NotaDuplicada extends Excecao{
        int matricula;
        String codigo;
        public NotaDuplicada(int matricula, String codigo){
            super("O aluno "+ matricula + "foi registrado em mais de um grupo para a avaliação "+codigo+".");
        }
    } 

    public static class IOException extends Excecao {
        public IOException() {
            super("Erro de I/O.");
        }
    }
}

File: TrabalhoC++/Excecao.cpp
#include "Excecao.hpp"

Excecao::Excecao(const std::string &msg) : std::runtime_error(msg) {}

FinalizaProgramaException::FinalizaProgramaException()
    : Excecao("") {}

ErroFormatacaoException::ErroFormatacaoException()
    : Excecao("Erro de formatação.") {}

CodigosIguaisException::CodigosIguaisException(int codigo)
    : Excecao("Código repetido para curso: " + std::to_string(codigo) + ".\n"),
      codigo(codigo) {}

MatriculasIguaisException::MatriculasIguaisException(int matricula)
    : Excecao("Matrícula repetida para aluno: " + std::to_string(matricula) + ".\n"),
      matricula(matricula) {}

CodDisciplinaIndefinidoAlunoExcpetion::CodDisciplinaIndefinidoAlunoExcpetion(int matricula, const std::string &codigo)
    : Excecao("Código de disciplina não definido usado no aluno " + std::to_string(matricula) + ": " + codigo + ".\n"),
      matricula(matricula),
      codigo(codigo) {}

CodDisciplinaIndefinidoAvalExcpetion::CodDisciplinaIndefinidoAvalExcpetion(const std::string &codigoA, const std::string &codigoD)
    : Excecao("Código de disciplina não definido usado na avaliação " + codigoA + ": " + codigoD + ".\n"),
      codigoA(codigoA),
      codigoD(codigoD) {}

PesoZeroNegativo::PesoZeroNegativo(const std::string &codigo, double peso)
    : Excecao("Peso de avaliação inválido para " + codigo + ": " + std::to_string((int)peso) + ".\n"),
      codigo(codigo),
      peso(peso) {}

NemPNemTException::NemPNemTException(const std::string &codigo, const std::string &digitado)
    : Excecao("Tipo de avaliação desconhecido para " + codigo + ": " + digitado + ".\n"),
      codigo(codigo),
      Digitado(digitado) {}

TamGrupoNaProvaException::TamGrupoNaProvaException(const std::string &codigo, int tam)
    : Excecao("Tamanho máximo de grupo especificado para a prova " + codigo + ": " + std::to_string(tam) + ".\n"),
      codigo(codigo),
      tam(tam) {}

TamMaxZeroNegativo::TamMaxZeroNegativo(const std::string &codigo, int peso)
    : Excecao("Tamanho máximo de grupo inválido para trabalho " + codigo + ": " + std::to_string(peso) + ".\n"),
      codigo(codigo),
      peso(peso) {}

NemGNemPException::NemGNemPException(int matricula, const std::string &digitado)
    : Excecao("Tipo de aluno desconhecido para " + std::to_string(matricula) + ": " + digitado + ".\n"),
      matricula(matricula),
      Digitado(digitado) {}

CodCursoIndefinidoException::CodCursoIndefinidoException(int matricula, int codigo)
    : Excecao("Código de curso não definido usado no aluno " + std::to_string(matricula) + ": " + std::to_string(codigo) + ".\n"),
      matricula(matricula),
      codigo(codigo) {}

NemMNemDException::NemMNemDException(int matricula, const std::string &digitado)
    : Excecao("Tipo de aluno de pós-graduação desconhecido para " + std::to_string(matricula) + ": " + digitado + ".\n"),
      matricula(matricula),
      Digitado(digitado) {}

CodAvaliacaoIndefinidoException::CodAvaliacaoIndefinidoException(const std::string &matriculas, const std::string &codigo)
    : Excecao("Código de avaliação não definido usado na planilha de notas, associado ao(s) aluno(s) " + matriculas + ": " + codigo + ".\n"),
      Matriculas(matriculas),
      Codigo(codigo) {}

MatriculaIndefinidaException::MatriculaIndefinidaException(const std::string &codigo, int matricula)
    : Excecao("Matricula de aluno não definida usada na planilha de notas, asssociada à avaliação " + codigo + ":" + std::to_string(matricula) + ".\n"),
      codigo(codigo),
      matricula(matricula) {}

NotaInvalidaAvaliacaoException::NotaInvalidaAvaliacaoException(const std::string &matriculas, const std::string &codigo, const std::string &nota)
    : Excecao("Nota inválida para avaliação " + codigo + " do(s) aluno(s) " + matriculas + ": " + nota + ".\n"),
      M(matriculas),
      codigo(codigo),
      nota(nota) {}

DisciplinaSemAvaliacaoException::DisciplinaSemAvaliacaoException(const std::string &codigo)
    : Excecao("A disciplina " + codigo + " não possui nenhuma avaliação cadastrada.\n"),
      codigo(codigo) {}

AlunoNaoMatriculadoException::AlunoNaoMatriculadoException(int matricula, const std::string &codigoA, const std::string &codigoD)
    : Excecao("O aluno " + std::to_string(matricula) + " possui nota na avaliação " + codigoA + " da disciplina " + codigoD + ", porém não encontra-se matriculado nesta disciplina.\n"),
      matricula(matricula),
      codigoA(codigoA),
      codigoD(codigoD) {}

NotaDuplicada::NotaDuplicada(int matricula, const std::string &codigo)
    : Excecao("O aluno " + std::to_string(matricula) + " foi registrado em mais de um grupo para a avaliação " + codigo + ".\n"),
      matricula(matricula),
      codigo(codigo) {}

ErroDeIO::ErroDeIO() : Excecao("Erro de IO\n") {}